pkg/pool-utils/contracts/RestrictedBalancerPoolToken.sol

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+
+import "./BalancerPoolToken.sol";
+import "./BasePoolAuthorization.sol";
+
+/**
+ * @title Extension of BalancerPoolToken with holder allowlist
+ * @author Balancer Labs
+ * @dev
+ * - Restricts both senders and recipients of all transfers (including mints
+ *   and burns) to addresses on an allowlist
+ * - Includes permissioned functions for adding/removing members to/from the
+ *   allowlist
+ */
+abstract contract RestrictedBalancerPoolToken is BalancerPoolToken, BasePoolAuthorization {
+    mapping(address => bool) private _allowedAddresses;
+
+    event AddressAdded(address member);
+    event AddressRemoved(address member);
+
+    constructor(string memory tokenName, string memory tokenSymbol, address owner)
+        BalancerPoolToken(tokenName, tokenSymbol)
+        BasePoolAuthorization(owner)
+    {
+        // Add zero address to allowlist to avoid blocking mints and burns
+        _allowedAddresses[address(0)] = true;
+    }
+
+    // Overrides
+
+    /**
+     * @dev Override to enforce address allowlist.
+     */
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256
+    ) internal virtual override {
+        _require(isAllowedAddress(from), Errors.ERC20_TRANSFER_FROM_PROHIBITED_ADDRESS);
```

Comment:
Do we need this? I imagined restricting the recipients would be enough - the only way a sender would not be allowed would be if at some point they were allowed, but that was later revoked.

Reply:
I think the situation you described is actually be a valid use case: someone is on the allowlist and then later removed because of bad behavior.